pkg/server/register.go

```diff
@@ -121,101 +119,135 @@ func upgrade(resp http.ResponseWriter, req *http.Request) (*websocket.Conn, erro
+func (i *InventoryServer) updateMachineInventory(inventory *elementalv1.MachineInventory) (*elementalv1.MachineInventory, error) {
+	if err := i.Update(i, inventory); err != nil {
+		return nil, fmt.Errorf("failed to update machine inventory %s/%s: %w", inventory.Namespace, inventory.Name, err)
 	}
-	return inventory, err
+
+	logrus.Infof("machine inventory updated: %s", inventory.Name)
+
+	return inventory, nil
 }
 
-func (i *InventoryServer) getMachineRegistration(req *http.Request) (*elm.MachineRegistration, error) {
+func (i *InventoryServer) getMachineRegistration(req *http.Request) (*elementalv1.MachineRegistration, error) {
 	token := path.Base(req.URL.Path)
 
-	regs, err := i.machineRegistrationCache.GetByIndex(registrationTokenIndex, token)
-	if apierrors.IsNotFound(err) || len(regs) != 1 {
-		if len(regs) > 1 {
-			logrus.Errorf("Multiple MachineRegistrations have the same token %s: %v", token, regs)
-		}
-		if err == nil && len(regs) == 0 {
-			err = fmt.Errorf("MachineRegistration does not exist")
+	mRegistrationList := &elementalv1.MachineRegistrationList{}
+	if err := i.List(i, mRegistrationList); err != nil {
+		return nil, fmt.Errorf("failed to list machine registrations")
+	}
+
+	var mRegistration *elementalv1.MachineRegistration
+
+	// TODO: build machine registrations cache indexed by token
+	for _, m := range mRegistrationList.Items {
+		if m.Status.RegistrationToken == token {
+			// Found two registrations with the same registration token
+			if mRegistration != nil {
+				return nil, fmt.Errorf("machine registrations %s/%s and %s/%s have the same registration token %s",
+					mRegistration.Namespace, mRegistration.Name, m.Namespace, m.Name, token)
+			}
+			mRegistration = &m
 		}
-		return nil, err
+	}
+
+	if mRegistration == nil {
+		return nil, fmt.Errorf("failed to find machine registration with registration token %s", token)
 	}
 
 	var ready bool
-	for _, condition := range regs[0].Status.Conditions {
+	for _, condition := range mRegistration.Status.Conditions {
 		if condition.Type == "Ready" && condition.Status == "True" {
 			ready = true
 			break
 		}
 	}
 
 	if !ready {
-		return nil, errors.New("MachineRegistration is not ready")
+		return nil, fmt.Errorf("MachineRegistration %s/%s is not ready", mRegistration.Namespace, mRegistration.Name)
 	}
 
-	return regs[0], nil
-
+	return mRegistration, nil
 }
 
-func (i *InventoryServer) writeMachineInventoryCloudConfig(conn *websocket.Conn, inventory *elm.MachineInventory, registration *elm.MachineRegistration) error {
-	var err error
+func (i *InventoryServer) writeMachineInventoryCloudConfig(conn *websocket.Conn, inventory *elementalv1.MachineInventory, registration *elementalv1.MachineRegistration) error {
+	var sa *corev1.ServiceAccount
+
+	err := i.Get(i, types.NamespacedName{
+		Name:      registration.Status.ServiceAccountRef.Name,
+		Namespace: registration.Status.ServiceAccountRef.Namespace,
+	}, sa)
```

Comment:
```suggestion
	if err := i.Get(i, types.NamespacedName{
		Name:      registration.Status.ServiceAccountRef.Name,
		Namespace: registration.Status.ServiceAccountRef.Namespace,
	}, sa); err != nil {
```

Reply:
The error is checked in the following line... but guess you that and you are hinting here that the right pattern for client runtime usage is the contract form `if err := i.Get(...`. Am I right?
Makes sense, will adhere to the format (if there is more just tell!).

Reply:
inlining `if err := ...`  is just idiomatic in golang :D